byway/src/main/webapp/routepage.js

```diff
@@ -49,27 +63,145 @@ function calcRoute(directionsService, directionsRenderer, start, end) {
+  clearStops();
+  fetch('/api/stop')
   .then(response => response.json())
-  .then((recs) => {
-    const recList = document.getElementById('rec-list');
-    recList.innerText = ""; // clear old comments 
-    recs.forEach((stop) => {
-      let btn = document.createElement("button");
-      btn.innerText = stop.placename;
-      btn.setAttribute("class", "btn rec-btn");
-      btn.addEventListener("click", function() {
-        addToStops(stop);
-      });
-      recList.appendChild(btn);
+  .then((stopsResponse) => {
+    stopsResponse.forEach((stop)=>{
+      renderStops(stop, true);
     });
+  })
+}
+
+/** Get the new list of stops locally */
+function getStopsList(){
+  clearStops();
+  // re-render list synchronously
+  stops.forEach((stop)=>{
+    renderStops(stop, false);
+  })
+}
+
+/** Add stop locally and to datastore */
+function addToStops(stop){
+  deleteFromRecs(stop); 
+  // add to stops array locally in js
+  stops.push(stop);
+  getStopsList();
+
+  // add to datastore
+  const params = new URLSearchParams();
+  params.append("text", stop.placename);
+  params.append("action", "add");
+  fetch('/api/stop', {method: 'POST', body: params});
+}
+
+/** Delete stop locally and from datastore*/
+function deleteFromStops(stop){
+  addToRecs(stop); 
+  console.log("delete " + stop.placename);
+  // delete from stops array locally in js
+  stops = stops.filter(function(stopObj){
+    return stopObj.placename != stop.placename;
+  })
+  getStopsList();
+
+  // delete from datastore
+  const params = new URLSearchParams();
+  params.append("text", stop.placename);
+  params.append("action", "remove");
+  fetch('/api/stop', {method: 'POST', body: params});
+}
+
+/** Clear the recommendations panel in the html */
+function clearRecs(){
+  const recList = document.getElementById('rec-list');
+  if(recList != null){
+    recList.innerText = ""; // clear list
+  }
+}
+
+/** Render recommendations list */
+function renderRecs(rec, fromDatastore){
+  const recsList = document.getElementById('rec-list');
+  // populate local js array initially
+  if(fromDatastore){
+    recs.push(rec);
+  }
+  var btn = document.createElement('button');
+  btn.id = rec.id;
+  btn.innerText = rec.placename;
+  btn.className =  "btn rec-btn";
+  btn.addEventListener("click", function() {
+    addToStops(rec); 
   });
+  recsList.appendChild(btn);
+}
+
+/** Get the new list of recommendations from servlet onload */
+function getRecsOnload() {
+  clearRecs();
+  fetch('/api/recs')
+  .then(response => response.json())
+  .then((recommendations) => {
+    recommendations.forEach((rec)=>{
+      renderRecs(rec, true);
+    })
+  })
+}
+
+/** Get the new list of recommendations locally */
+function getRecsList(){
+  clearRecs();
+  // re-render list synchronously
+  recs.forEach((rec)=>{
+      renderRecs(rec, false);
+  })
+}
+
+/** Add recommendation locally*/
+function addToRecs(rec){
+  recs.push(rec);
+  getRecsList();
+  const params = new URLSearchParams();
```

Comment:
What is this used for? Did you mean to delete these two lines?

Reply:
Oh yes, that's a mistake. I originally was changing the recommendations in the datastore, but you suggested not removing anything from that. Right now, if you refresh the page you get the same hardcoded places rendered as there is no deletion/addition. but the schedule/selected stops panel does get updated properly. 

Also there's a lot of uncertainty with getting the recommendations into datastore that I will address in a future pull request.